### PR TITLE
Encoder multiplier - revised code

### PR DESCRIPTION
**Pull request by thinkyhead**
Tweaks to spacing

### CODE REVIEW

**Comment by alexborro** [comment]
What is the encoder multiplier meant for?

**Comment by thinkyhead** [comment]
@alexborro This stems from comments about UX to reduce the amount of knob-turning required to reach a desired value (#1518 and #1411). This change allows adjustable menu settings that opt-in to employ acceleration. That is, if controller rotation speed goes over a certain threshold, it produces a higher rate of change. It's a good idea, and should get some UX testing with different controllers to see that it feels proper.

The behavior is, I think, that the rate of change goes up, so values will proceed (slow, fast, slow) 1, 2, 3 … 13, 23, 33 … 34, 35, 36 … Another option would be to have values proceed (slow, fast, slow) 1, 2, 3 … 10, 20, 30 … 31, 32, 33 … The feel is improved if, when values go up by larger increments, the delay between ticks is slightly longer. (Perhaps that is how it is - will test soon.)

**Comment by alexborro** [comment]
That's is nice.. The encoder operation is something that bores me a lot!!

In my LCD 2004 and Full Graphic LCD, the encoder needs 2 "ticks" to move
the cursor once and then more 3 "ticks" for the next movement. I don't know
why that happens.. 2 ticks, 3 ticks, 2 ticks, 3 ticks.
That's really odd..

Does this PR fix it??

2015-02-20 10:07 GMT-02:00 Scott Lahteine notifications@github.com:

> @alexborro https://github.com/alexborro This stems from comments about
> UX to reduce the amount of knob-turning required to reach a desired value (
> #1518 https://github.com/MarlinFirmware/Marlin/pull/1518 and #1411
> https://github.com/MarlinFirmware/Marlin/issues/1411). This change
> allows adjustable menu settings that opt-in to employ acceleration. That
> is, if controller rotation speed goes over a certain threshold, it produces
> a higher rate of change. It's a good idea, and should get some UX testing
> with different controllers to see that it feels proper.
> 
> The behavior is, I think, that the rate of change goes up, so values will
> proceed (slow, fast, slow) 1, 2, 3 … 13, 23, 33 … 34, 35, 36 … Another
> option would be to have values proceed (slow, fast, slow) 1, 2, 3 … 10, 20,
> 30 … 31, 32, 33 … The feel would be improved if, when values go up by
> larger increments, that the delay between ticks is slightly longer.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/MarlinFirmware/Marlin/pull/1521#issuecomment-75228555
> .

## 

_"Não é o mais forte da espécie que sobrevive, nem o mais inteligente. É
aquele que se adapta melhor as mudanças"  ( Charles Darwin )_

Alex Borro

**Comment by wgm4321** [comment]
@thinkyhead and @alexborro it's not as sophisticated as that.... Basically if you "spin" fast, it multiplies the steps that are reported to the code doing the value edit.   if you spin REAL fast, it's 100x.

So for example, if you go to the fan speed, a quick half turn snap spin gets you to 255 due to the 100x.   and a quick half turn snap the other way will get you back to zero.    However, if you slow down some, the same half turn might get you to 90.    Slow down even more and the same half turn gets you the normal amount of steps..

The rates of "spin" that trigger the 10x and 100x are in configuration_adv.h.   There is also a define that will output the rate and what multiplier was used for the movements that might help you understand and/or dial in the rate levels.

We can try different algorithms but this was the simplest and has the  least dependencies on timing.   Trying to do something like "as long as the knob is turned at a certain rate for a certain time" might "feel" better but tracking that precisely could be difficult if the sequence is interrupted by moving to a new menu etc.

Let me know what you think.   Note that only certain edit functions enable this feature.   The filament size, hotend temps, bed temps and fan speed.   If there are other values that would benefit from this, you just change the MENU_ITEM_EDIT macro to MENU_MULTIPLIER_ITEM_EDIT for the menu item you want this enabled on.    If the multiplier function is disabled in configuration_adv.h, MENU_MULTIPLIER_ITEM_EDIT is simple defined to be MENU_ITEM_EDIT so you can disable and not have to change all the menu item definitions if this breaks something .

@thinkyhead sorry about the spacing.   I'm using visual studio (stop laughing, it's pretty good with the exception of the following) and it likes to reformat spacing and tabs....   Did you introduce the missing slash or did I?   I'd like to know when I screw up so I don't do it again....

**Comment by wgm4321** [comment]
@thinkyhead I'm new to git but I see that this is separate from the pull request I made https://github.com/MarlinFirmware/Marlin/pull/1518
Do I need to do something to "cancel" that request?

**Comment by thinkyhead** [comment]
@wgm4321 I honestly am not sure about this PR's relationship to your PR. What I wanted to do was clone your PR's _branch_, make changes, commit, publish, _and then_ make a PR that you would receive _on your fork's branch_ that you could then choose to merge, or not, and it would then end up as a new commit in your original PR.

I couldn't get that worked out, so I ended up with a clone of your branch, plus my changes, as a new branch and a new PR. I'd like to figure out how to do this kind of thing really simply. I think it would take less time to learn what git _cannot_ do than what it can.... I think I needed to make your fork into a new remote, then I could fetch from your remote.... oh my head....

Anyway, I like the simpler approach than what I described, where faster movement just produces some proportional speedup – it's no different from _mouse acceleration_. The other concept makes a certain sense, as for example we want to tune temperatures in 5 degree increments (generally, then fine-tune).

**Comment by wgm4321** [comment]
No problem.   Let's just use this for any changes and I'll see if I can "cancel" my other PR.
GIT is pretty amazing as far as features go....   but that makes the learning curve longer.

**Comment by thinkyhead** [comment]
Cleaned up & ready to go. Includes a semi-related bugfix, so will merge pretty soon.

File: Marlin/ultralcd.cpp [hunk 1]
```diff
@@ -841,28 +876,28 @@ static void lcd_control_temperature_menu()
...
-      MENU_ITEM_EDIT(int3, MSG_NOZZLE " 2", &target_temperature[1], 0, HEATER_1_MAXTEMP - 15);
+      MENU_MULTIPLIER_ITEM_EDIT(int3, MSG_NOZZLE " 2", &target_temperature[1], 0, HEATER_1_MAXTEMP - 15);
     #endif
     #if EXTRUDERS > 2
       #if TEMP_SENSOR_2 != 0
-        MENU_ITEM_EDIT(int3, MSG_NOZZLE " 3", &target_temperature[2], 0, HEATER_2_MAXTEMP - 15);
+        MENU_MULTIPLIER_ITEM_EDIT(int3, MSG_NOZZLE " 3", &target_temperature[2], 0, HEATER_2_MAXTEMP - 15);
       #endif
-      #if EXTRUDERS > 2
+      #if EXTRUDERS > 3
```
Comment (thinkyhead) [inline Marlin/ultralcd.cpp:889]: Bugfix!

File: Marlin/ultralcd.cpp [hunk 2]
```diff
@@ -828,7 +863,7 @@ static void lcd_control_menu()
         void copy_and_scalePID_d_E3() { copy_and_scalePID_d(2); }
         #if EXTRUDERS > 3
           void copy_and_scalePID_i_E4() { copy_and_scalePID_i(3); }
-          void copy_and_scalePID_d_E5() { copy_and_scalePID_d(3); }
+          void copy_and_scalePID_d_E4() { copy_and_scalePID_d(3); }
```
Comment (thinkyhead) [inline Marlin/ultralcd.cpp:866]: Bugfix!

**Comment by thinkyhead** [comment]
Also want to get this in ahead of #1526 …

**Pull request merged**